Support for LedgerState and folding over all blocks

DavidEichmann opened the pull request:
The `foldBlocks` function was originally used by @JaredCorduan to implement an audit tool. In my first commit I remove the existing cardano-api `LedgerState` type and corresponding queries. This is to make room for the new `LedgerState` type that I introduce in the second commit.

I eventually plan to add a query for the full ledger state. Once that is merged, I'll be able to reintroduce the queries that I've removed here.

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,913 @@
+    ncfg <- parseNodeConfig <$> readByteString ncf "node"
+    coalesceConfig ncfg (mkAdjustPath ncf)
+
+data NodeConfig = NodeConfig
```
There is `NodeConfiguration` in `cardano-node` that has this information and can be extended if there are fields missing.

DavidEichmann replied:
I think I misunderstood this yesterday. Seeing as I'll be able to remove this once the query for ledger config gets merged, Perhaps we're ok with having `NodeConfig` for now? I'll add a TODO in the code.

Jimbo4350 replied:
No worries. I think the best solution would be to move the node's configuration stuff to the api. This is good for now.

Jimbo4350 replied:
Update: The genesis file will eventually not be needed so we don't need to move the node configuration stuff to cardano-api

Jimbo4350 reviewed (commented):
Some initial comments. This looks like it was pretty hairy to sort out :sweat_smile:

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,913 @@
+  -- ^ The environment and initial ledger state
+initialLedgerState dbSyncConfFilePath = do
+  dbSyncConf <- readDbSyncNodeConfig (NodeConfigFile dbSyncConfFilePath)
+  genConf <- fmap (either (error . Text.unpack . renderDbSyncNodeError) id) $ runExceptT (readCardanoGenesisConfig dbSyncConf)
```
If this is eventually intended for `cardano-cli` let's return an `Either` so we can handle errors there.

DavidEichmann replied:
Yes... actually I used `error` in a few places and need to turn these into propper errors

DavidEichmann replied:
@Jimbo4350 I'm using `ExceptT` now. Is that ok with you?

DavidEichmann commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,913 @@
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
```
That and other extensions are not needed. I'll remove them.

DavidEichmann commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,913 @@
+  connectInfo =
+      LocalNodeConnectInfo {
+        localConsensusModeParams = CardanoModeParams (EpochSlots 21600),
+        localNodeNetworkId       = Mainnet,
```
I'll turn it into a parameter

DavidEichmann commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,913 @@
+        clientIdle_RequestMoreN
+          :: WithOrigin BlockNo
+          -> WithOrigin BlockNo
+          -> Nat n
```
"Number of requests inflight." will add a comment

DavidEichmann commented:
Changes:
* Use `ExceptT` instead of `error`
* Remove `DbSyncNodeConfig` and use `NodeConfig` instead. There were isomorphic any way (just an artifact of copying from db-sync)
* Actually check the Shelley hash from the config file.
* Take the network as a parameter to `foldBlocks`

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,849 @@
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -freduction-depth=0 #-}
```
I can seem to find any docs on this. Do you have a source?

DavidEichmann replied:
Wow.. not sure how that got there. I've removed it.

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,849 @@
+    "Failed to read or parse a genesis file linked from the network config file: "
+    <> renderGenesisConfigError err
+  ILSELedgerConsensusConfig err ->
+    "Failed to derived the Ledger or Consensus config: "
```
should be derive

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,849 @@
+  | ILSEGenesisFile GenesisConfigError
+  -- ^ Failed to read or parse a genesis file linked from the network config file.
+  | ILSELedgerConsensusConfig GenesisConfigError
+  -- ^ Failed to derived the Ledger or Consensus config.
```
should be derive

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,849 @@
+    Just err -> throwE (FoldBlocksApplyBlockError err)
+    Nothing -> lift $ readIORef stateIORef
+  where
+  connectInfo :: LocalNodeConnectInfo CardanoMode
```
Lets put an indent here to align with the coding style of the repo. This will make the `where` clause stand out more.

Jimbo4350 commented on cardano-api/src/Cardano/Api/LedgerState.hs:
```diff
@@ -0,0 +1,849 @@
+      }
+
+  -- | Add a new ledger state to the history
+  pushLedgerState :: Env -> LedgerStateHistory -> SlotNo -> LedgerState -> BlockInMode CardanoMode
```
Annotations on these types would be useful as well.

Jimbo4350 commented on cardano-api/src/Cardano/Api/Query.hs:
```diff
@@ -148,9 +144,6 @@ data QueryInShelleyBasedEra era result where
      -- QueryPoolRanking
      --   :: QueryInShelleyBasedEra era RewardProvenance
 
-     QueryLedgerState
```
Why are we removing the `QueryLedgerState`? Can't we leave it for now until you're ready to replace it?

Jimbo4350 reviewed (commented):
Good work, the code is a lot clearer now. A couple comments but nothing major.

DavidEichmann commented:
Update:

* The first commit simply renames some things instead of deleting the cli ledger state query
* Resolved PR comments

Jimbo4350 reviewed (approved):
LGTM! Is there a TODO regarding potentially moving the `NodeConfiguration` machinery to `cardano-api`? You should also update the PR description as you are no longer removing a query.

DavidEichmann commented:
> Is there a TODO regarding potentially moving the NodeConfiguration machinery to cardano-api?
I added this: https://github.com/input-output-hk/cardano-node/pull/2633/files#diff-b415c60c9e90bf9d54569de2fb6117807512a0f98488005a076913cafa7c947dR133

Update:
* Removed an old TODO and cleaned up some imports

DavidEichmann commented:
Update:
* Rebase
* ouroboros-network (version bumped after the rebase) no longer exports `Protocol` which this code used transitively. Removing use of `Protocol` turned out to be pretty trivial (after a bit of fumbling).

DavidEichmann commented:
bors merge

DavidEichmann commented:
Update:
* Fix hlint errors

DavidEichmann commented:
bors merge

JaredCorduan commented:
:tada: `foldBlocks` fills me with joy, thanks again @DavidEichmann !
